### PR TITLE
[Hack-373] CircleCI Orb to Publish Github Releases

### PR DESCRIPTION
**Pull request by uttercm**
**JIRA Ticket:**
[HACK-373](https://civisanalytics.atlassian.net/browse/HACK-373)

**Summary:**
We want a CircleCI orb that can publish github releases and potentially publish releases to PyPi. This is the first iteration of this Orb to be released as version `civisanalytics/github-release@dev:alpha` It has one command and two jobs. One job that can take a version and changelog file and create a release based on parsing those files for the version and description.

**Testing:**
- [x] Performed manual tests
I tested the jobs and command within the JDBCTool repository and a python repository (identity-resolution).

Notes:

**Additional Info:**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,87 @@
...
+  steps:
+    - github-release/<command>
+```
+
+## Build
+In order to build the orb you can run the following command in the root directory of this repository.
+
+`circleci config pack src > orb.yml`
+
+Do not commit the created `orb.yml` file. This file is for publishing to CircleCI when finished.
```
Comment (zacharydfriedlander) [inline README.md]: If this should not be committed, can we add it to a `.gitignore`?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,87 @@
...
+| `title` | string | Name of the release | |
+| `description` | string | Description of the release | |
+| `file-path` | string | File directory to be uploaded in addition to the full repository. | `""`|
+| `is-draft` | boolean | If true, then the release is a draft. | `false` |
+| `is-pre-release` | boolean | If true, then the release is a pre-release. | `false` |
+
+*No default value means the parameter is implied as required.*
+
+### release-by-version
+This is a job to publish a new GitHub Release by checking the version file and CHANGELOG files provided. This is to help ease the process of updating the description for the automated releases. It's intended repository use is for Python since the default version-path is set to `version.py`
```
Comment (zacharydfriedlander) [inline README.md]: We should have a period at the end of this sentence.

Comment (zacharydfriedlander) [reply README.md]: This last sentence seems a little strangely worded to me; It sounds like default of the parameter is causing the intended use case, rather than the other way around! What if we phrased this as something like:

> Since this orb is intended for use with Python repositories, the `version-path` parameter defaults to `version.py`.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,87 @@
...
+| Parameter | Type | Description | Default |
+| --- | --- | --- | --- |
+| `current-version` | string | Current tagged version of the new release | |
+| `title` | string | Name of the release | |
+| `description` | string | Description of the release | |
+| `file-path` | string | File directory to be uploaded in addition to the full repository. | `""`|
+| `is-draft` | boolean | If true, then the release is a draft. | `false` |
+| `is-pre-release` | boolean | If true, then the release is a pre-release. | `false` |
+
+*No default value means the parameter is implied as required.*
```
Comment (zacharydfriedlander) [inline README.md]: I would rather add something to the table itself than have this disclaimer; I can think of two options:

1. Add a `Required?` column to the table between `Description` and `Default`.
2. Add an asterisk or some other indicator to the `Parameter` column for required parameters.

File: src/commands/publish.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      name: Provide error if non-bash shell
+      command: |
+        if [ ! -x /bin/bash ]; then
+          echo Bash not installed.
+          exit 1
+        fi
+  - run:
+      name: Provide error if cURL is not installed.
+      command: |
+        which curl > curl_exists; echo $? | grep -q '1' && echo curl not installed && rm curl_exists && exit 1
```
Comment (zacharydfriedlander) [inline src/commands/publish.yml]: Why does this need to go to a file? Can we not just `which curl` and then check `$?` without a file? 

Alternatively, if we _do_ need to send output somewhere, we should send it to [`/dev/null`](https://en.wikipedia.org/wiki/Null_device); that way we don't have a file that needs to get removed, but we still don't have output.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,87 @@
...
+## Dependencies / Requirements
+We have checks in place to make sure the following are installed within the container otherwise an error message will be logged and task failed.
+
+### Bash Shell
+These commands require `Bash` in order to function. `Bash` is the default shell for CircleCI; therfore, this Orb should be compatible with most images.
+
+### cURL
+cURL is used to get TAG data for the GitHub releases and must be installed in the container to function.
+
+### awk
```
Comment (zacharydfriedlander) [inline README.md]: So if we need these... can we just install them ourselves? Or is that an orb no-no?

File: src/examples/publish.yml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+description: Publish a new GitHub Release.
```
Comment (zacharydfriedlander) [inline src/examples/publish.yml:1]: ⭐️ for having these examples!

File: src/jobs/release-by-version.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  is-pre-release:
+    description: If true, then the release is a pre-release.
+    type: boolean
+    default: false
+
+steps:
+  - checkout
+  - run:
+      name: Provide error if awk is not installed.
+      command: |
```
Comment (zacharydfriedlander) [inline src/jobs/release-by-version.yml]: Same as above - do we need a file? If we do, can we use `/dev/null` instead?

**Review by zacharydfriedlander — CHANGES_REQUESTED** [review]
A few thoughts:

File: src/commands/publish.yml (hunk 4 above)
Comment (uttercm) [reply src/commands/publish.yml]: no reason, can change

File: src/jobs/release-by-version.yml (hunk 7 above)
Comment (uttercm) [reply src/jobs/release-by-version.yml]: nope

File: README.md (hunk 5 above)
Comment (uttercm) [reply README.md]: for the `commands`, I don't think we want to be installing them in the docker image if they aren't provided, but for the `jobs` within the orb, we use the `cibuilds/github` which uses the `cibuilds/base` that installs the dependencies so really we don't need the `awk` check within that.

Comment (uttercm) [reply README.md]: i'm going to add in the dependencies section the docker images that have these built in

File: src/commands/publish.yml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  - run:
+      name: Publish Release on GitHub
+      shell: /bin/bash
+      command: |
+        URL="https://api.github.com/repos/${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}/releases/latest"
+        LAST_TAG=$(curl -s -H "Authorization: token ${GITHUB_TOKEN}" $URL | jq -r .tag_name)
+        LAST_VERSION=${LAST_TAG#'v'}
+        VERSION="<< parameters.current-version>>"
+        if [ $VERSION != $LAST_VERSION ]; then
+          ghr \
```
Comment (uttercm) [inline src/commands/publish.yml:54]: don't need to check this command like the others since this will actually error as expected

File: README.md [hunk 9]
```diff
@@ -1,2 +1,89 @@
...
+| Parameter | Type | Description | Required | Default |
+| --- | --- | --- | --- | --- |
+| `version-path` | string | File path to file containing the version. If not specified will default to version.py. We will split the line matching version=x.y.z to get the version. | false |  `version.py`|
+| `changelog-path` | string | File path to the changelog containing the description of the release. | false | `CHANGELOG.md` |
+| `file-path` | string | File directory to be uploaded in addition to the full repository. | false | `""`|
+| `is-draft` | boolean | If true, then the release is a draft. | false | `false` |
+| `is-pre-release` | boolean | If true, then the release is a pre-release. | false |`false` |
+
+## Dependencies / Requirements
+We have checks in place to make sure the following are installed within the container otherwise an error message will be logged and task failed. If your job uses, the `github` executor, these dependencies will be installed. Otherwise, we recommend the following Docker images that will ensure these are downloaded:
```
Comment (zacharydfriedlander) [inline README.md]: I don't think there should be a comma after `uses` in the second sentence here:

> If your job uses the `github` executor,...

File: src/commands/publish.yml [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  is-pre-release:
+    description: If true, then the release is a pre-release.
+    type: boolean
+    default: false
+
+steps:
+  - run:
+      name: Provide error if non-bash shell
+      command: |
+        if [ ! -x /bin/bash ]; then
```
Comment (zacharydfriedlander) [inline src/commands/publish.yml:36]: I'm not sure this is the command we want to be using; this just checks that bash is installed ,not whether or not we're using it here. It seems like we can do something [like this](https://askubuntu.com/questions/590899/how-do-i-check-which-shell-i-am-using), combined with [parameter expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) to check that the current shell is `bash`:

```
if [ ${SHELL##/*/} != 'bash' ]
```

**Review by zacharydfriedlander — CHANGES_REQUESTED** [review]
A few more thoughts:

File: src/commands/publish.yml (hunk 10 above)
Comment (uttercm) [reply src/commands/publish.yml:36]: I don't think this works, CircleCI must do something odd with those parameters, I get nothing for the `SHELL` value and `echo $0` returns `ash`

Comment (uttercm) [reply src/commands/publish.yml:36]: I think it might be fine that we are only checking that bash is installed since line 47 requires us to use it.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,89 @@
...
+
+`circleci config pack src > orb.yml`
+
+Do not commit the created `orb.yml` file. This file is for publishing to CircleCI when finished.
+
+## Required Environment Variables
+In order for this orb to work, you must have three Environment Variables set:
+1. `GITHUB_TOKEN` - Your GitHub personal access token.
+2. `CIRCLE_PROJECT_USERNAME` - GitHub repository's user or organization. This should be set automatically.
+3. `CIRCLE_PROJECT_REPONAME` - GitHub repository name. This should be set automatically.
```
Comment (leanne73) [inline README.md:32]: I'm a little confused about this section - is it for development purposes or for actual usage? If for development, it might be nice to have a section dedicated to development instructions (It seems like the `Build` section is also related to development.)

I guess I also don't know much about orbs and how they work, maybe a brief intro would be nice? Even just something along the lines of "Use this orb to help automate your github releases via CircleCI"

File: README.md [hunk 12]
```diff
@@ -1,2 +1,89 @@
...
+| `is-draft` | boolean | If true, then the release is a draft. | false | `false` |
+| `is-pre-release` | boolean | If true, then the release is a pre-release. | false| `false` |
+
+### release-by-version
+This is a job to publish a new GitHub Release by checking the version file and CHANGELOG files provided. This is to help ease the process of updating the description for the automated releases. Since this orb is intended for use with Python repositories, the `version-path` parameter defaults to `version.py`.
+
+| Parameter | Type | Description | Required | Default |
+| --- | --- | --- | --- | --- |
+| `version-path` | string | File path to file containing the version. If not specified will default to version.py. We will split the line matching version=x.y.z to get the version. | false |  `version.py`|
+| `changelog-path` | string | File path to the changelog containing the description of the release. | false | `CHANGELOG.md` |
```
Comment (leanne73) [inline README.md]: Is this file expected to hold only the description for this release? It looks like we are parsing it in some way, is there an expected file format?

File: src/@orb.yml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+version: 2.1
```
Comment (leanne73) [inline src/@orb.yml:1]: Is this the orb's version? Why are we starting at 2.1?

File: src/@orb.yml [hunk 14]
```diff
@@ -0,0 +1,5 @@
...
+
+description: >
+  Create automated releases to GitHub for your CircleCI projects.
+  View this orb's source: https://github.com/civisanalytics/github-release-orb
```
Comment (leanne73) [inline src/@orb.yml:5]: Okay, more ignorance here: How are orbs published? Is this a public orb (& a public repo)? Or internal to Civis only?

File: src/commands/publish.yml [hunk 15]
```diff
@@ -0,0 +1,65 @@
+description: |
+  Publish a new Github release by checking the previous version with the given current version.
```
Comment (leanne73) [inline src/commands/publish.yml]: This is a little confusing to me, do we mean "Publish a new Github release if the previous version does not match the given current version"?

File: src/examples/publish.yml [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+  orbs:
+    github-release: civisanalytics/github-release@x.y.z
+  workflows:
+    your-workflow:
+      jobs:
+        - publish-github-release:
+            executor: github-release/github
+            steps:
+              - github-release/publish:
+                  current-version: a.b.c
```
Comment (leanne73) [inline src/examples/publish.yml]: I'm a little confused as to why you'd want to hard-code the version into your circle config... Ohh, is it that you'd use this workflow when you're using circleci config itself to track the version of your code (and then publish new releases if and only if it changes?)

File: src/examples/publish.yml (hunk 6 above)
Comment (leanne73) [reply src/examples/publish.yml:1]: I agree - these look great! Should there also be an example of using `publish-github-release`?

File: README.md [hunk 17]
```diff
@@ -1,2 +1,89 @@
...
+  steps:
+    - github-release/<command>
+```
+
+## Build
+In order to build the orb you can run the following command in the root directory of this repository.
+
+`circleci config pack src > orb.yml`
+
+Do not commit the created `orb.yml` file. This file is for publishing to CircleCI when finished.
```
Comment (leanne73) [inline README.md]: Should there be instructions on how to publish new versions of the orb? Or a link to docs on how to publish, if there are circle maintained docs

File: README.md [hunk 18]
```diff
@@ -1,2 +1,89 @@
...
+In order for this orb to work, you must have three Environment Variables set:
+1. `GITHUB_TOKEN` - Your GitHub personal access token.
+2. `CIRCLE_PROJECT_USERNAME` - GitHub repository's user or organization. This should be set automatically.
+3. `CIRCLE_PROJECT_REPONAME` - GitHub repository name. This should be set automatically.
+
+## Commands
+
+### Publish
+Publish a new GitHub Release if the last version does not match the current version. This command is useful for running within a custom job. If no custom job is needed, we recommend using the built-in `publish-github-release` job provided by this orb.
+
```
Comment (leanne73) [inline README.md:38]: Can you give an example of when a custom job would be useful? At least just for my own sake. Would this allow you to automate finding the current-version, for example?

File: README.md [hunk 19]
```diff
@@ -1,2 +1,89 @@
...
+| Parameter | Type | Description | Required | Default |
+| --- | --- | --- | --- | --- |
+| `current-version` | string | Current tagged version of the new release | true | |
```
Comment (leanne73) [inline README.md]: I'm a little confused by this variable/description. Should the variable just be `version`? Based on the rest of the code, it seems like this variable describes the most recently released version. That it describes the current version when not changing but the new version when changing. Would an appropriate description be `Version of this codebase. Will create a new github release if different from the previous version.`?

**Review by leanne73 — COMMENTED** [review]
I think that this looks really cool and its obvious that you've put a lot of work into making the documentation look nice! I've never worked with circleci orbs before, so I hope my questions aren't too ignorant!

File: src/examples/publish.yml (hunk 6 above)
Comment (uttercm) [reply src/examples/publish.yml:1]: can add

File: README.md (hunk 11 above)
Comment (uttercm) [reply README.md:32]: this is for actual usage, I can make it more clear since `Build` is for development

File: README.md (hunk 12 above)
Comment (uttercm) [reply README.md]: it's supposed to be in the format of a typical `CHANGELOG`, I'll add a link keepachangelog.com

File: src/@orb.yml (hunk 13 above)
Comment (uttercm) [reply src/@orb.yml:1]: this is the CircleCI version

File: src/@orb.yml (hunk 14 above)
Comment (uttercm) [reply src/@orb.yml:5]: I can add a  description in the Readme on how to publish the orb, all orbs are public, and this is a public repo.

File: src/examples/publish.yml (hunk 16 above)
Comment (uttercm) [reply src/examples/publish.yml]: ya, my thought is maybe someone doesn't have a version file and instead would want to publish versions through the config.

File: README.md (hunk 17 above)
Comment (uttercm) [reply README.md]: I can add those

File: README.md (hunk 18 above)
Comment (uttercm) [reply README.md:38]: maybe you want a custom job to first create a file to upload with the release that doesn't exist yet, kinda thinking along the lines of adding the JAR for the JDBCTool, then you use the created file-path in the command.

File: README.md (hunk 19 above)
Comment (uttercm) [reply README.md]: Ya that makes sense

File: src/commands/publish.yml (hunk 15 above)
Comment (uttercm) [reply src/commands/publish.yml]: say someone merged to master but didn't want to publish a new release we check that the version is not the same as the last tagged version

Comment (uttercm) [reply src/commands/publish.yml]: but yes I can change that to be more clear

File: README.md (hunk 18 above)
Comment (uttercm) [reply README.md:38]: i added to the examples a `custom-job`

File: LICENSE.md [hunk 20]
```diff
@@ -0,0 +1,9 @@
+Copyright 2019 Circle Internet Services, Inc.
```
Comment (zacharydfriedlander) [inline LICENSE.md]: 1. We should check with legal before we put any license on anything. This looks like the MIT license, which historically I think we've been okay with, but we should do our due diligence.
2. Why are we granting copyright to circle? Why not us?

File: README.md [hunk 21]
```diff
@@ -1,2 +1,120 @@
...
+| `version-path` | string | File path to file containing the version. If not specified will default to version.py. We will split the line matching `version=x.y.z` to get the version. | false |  `version.py`|
+| `changelog-path` | string | File path to the changelog containing the description of the release. The format should follow typical [Changlog Symantics](https://keepachangelog.com/en/1.0.0/). | false | `CHANGELOG.md` |
+| `file-path` | string | File directory to be uploaded in addition to the full repository. | false | `""`|
+| `is-draft` | boolean | If true, then the release is a draft. | false | `false` |
+| `is-pre-release` | boolean | If true, then the release is a pre-release. | false |`false` |
+
+### Dependencies / Requirements
+We have checks in place to make sure the following are installed within the container otherwise an error message will be logged and task failed. If your job uses the `github` executor, these dependencies will be installed. Otherwise, we recommend the following Docker images that will ensure these are downloaded:
+
+* cibuilds/github:0.13
```
Comment (zacharydfriedlander) [inline README.md]: Can we put a link in here to more info about this image?

File: README.md [hunk 22]
```diff
@@ -1,2 +1,120 @@
...
+#### Prerequisites
+Make sure you have CircleCI CLI installed on your machine. If you don't, you can follow [Setting Up the CircleCI CLI](https://circleci.com/docs/2.0/orb-author-cli/#section=configuration) to get setup.
+
+### Building and Publishing the Orb
+Once ready to publish a new version of the orb you can follow these steps:
+
+1. Build the orb by running: `circleci config pack src > orb.yml`. Do not commit the created `orb.yml` file. This file is for publishing to CircleCI when finished.
+2. Once the orb is built, we can verify the orb is valid by running `circleci orb validate orb.yml`.
+3. Publish a dev version of the orb by running `circleci orb publish orb.yml civisanalytics/github-release@dev:<version>`. This dev version is available only to civisanalytics and will expire 90 days after each publish. This can be used for testing the orb before a final release.
+4. When ready to publish the orb, email IT at it-help@civisanalytics.com to have them run `circleci orb publish promote civisanalytics/github-release@dev:<version> <segment>` where `segment` is either `major`, `minor`, or `patch` and it will auto-increment the published version of the orb.
```
Comment (zacharydfriedlander) [inline README.md]: If this is going to be public, do we want to put our internal IT helpdesk email in the Readme?

**Review by zacharydfriedlander — CHANGES_REQUESTED** [review]
A few more thoughts:

File: LICENSE.md (hunk 20 above)
Comment (uttercm) [reply LICENSE.md]: Ya, I can check with them, I didn't know what we should have in the license but I figured we do need a LICENSE file since this is a public repo.

File: README.md (hunk 22 above)
Comment (uttercm) [reply README.md]: hmm should it be something like `when ready to publish the orb, contact a Civis employee`

Comment (uttercm) [reply README.md]: also only IT members can publish an orb not all Civis employees

Comment (zacharydfriedlander) [reply README.md]: I'm not sure I like that phrasing either. I think I want to just say something like "the orb owner can run `circleci orb publish promote civisanalytics/github-release@dev:<version> <segment>`. That way, we don't mention us/IT at all.

File: LICENSE.md (hunk 20 above)
Comment (uttercm) [reply LICENSE.md]: I'm going to delete this for now until I hear back so we can get the first iteration of the orb  out.

File: src/@orb.yml (hunk 13 above)
Comment (leanne73) [reply src/@orb.yml:1]: ahhh, lol that makes sense

File: README.md [hunk 23]
```diff
@@ -1,2 +1,120 @@
...
+#### cURL
+cURL is used to get TAG data for the GitHub releases and must be installed in the container to function.
+
+#### ghr
+The tool that does the creation of the GitHub Release.
+
+#### awk
+Only the `release-by-version` is dependent upon `awk` to function and it is already installed with the executor for this job. We use this in order to parse the version and changelog files.
+
+## Development
```
Comment (leanne73) [inline README.md:107]: Yay! This section is super useful

File: src/commands/publish.yml [hunk 24]
```diff
@@ -0,0 +1,65 @@
+description: |
+  Publish a new Github release if the previous version does not match the given version.
+  Required environment variables for this Orb are:
+    1. GITHUB_TOKEN - Your GitHub personal access token
+    2. CIRCLE_PROJECT_USERNAME - GitHub repository's user or organization
+    3. CIRCLE_PROJECT_REPONAME - GitHub repository name
+
+parameters:
+  version:
+    description: Current tagged version of the new release
```
Comment (leanne73) [inline src/commands/publish.yml]: I think this needs updating, should probably match the new docs? e.g. "Version of the codebase. Orb will create a new github release only if different from the previous version."

File: src/examples/custom-job.yml [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+        - github-release/set-release-details
+        - github-release/publish:
+            version: $CURRENT_VERSION
+            title: $TITLE
+            description: $DESCRIPTION
+            file-path: $FILE
+  workflows:
+    your-workflow:
+      jobs:
+        - custom-job
```
Comment (leanne73) [inline src/examples/custom-job.yml:28]: This is a great example! I like that it shows how to use `set-release-details` too

File: src/jobs/publish-github-release.yml [hunk 26]
```diff
@@ -0,0 +1,36 @@
+description: Publish a new GitHub Release.
+
+executor: github
+
+parameters:
+  version:
+    description: Current tagged version of the new release
```
Comment (leanne73) [inline src/jobs/publish-github-release.yml]: I think this description needs updating too

File: src/jobs/release-by-version.yml [hunk 27]
```diff
@@ -0,0 +1,45 @@
+description: |
+  Publish a new GitHub Release for a repository by finding the version file and CHANGELOG then publishing a
+  release if the found version does not match the previous version.
```
Comment (leanne73) [inline src/jobs/release-by-version.yml]: Hmm. I find this a bit hard to follow, there's a lot of pieces here. Maybe: 
"Conditionally publish a new GitHub Release for a repository based on the version file. If the version in the version file does not match the latest Github Release, a new release will be generated. Release description will be parsed from CHANGELOG.md by default." ?

**Review by leanne73 — COMMENTED** [review]
This is looking good!

File: src/commands/publish.yml (hunk 24 above)
Comment (uttercm) [reply src/commands/publish.yml]: ah yes, thank you

File: src/jobs/release-by-version.yml (hunk 27 above)
Comment (uttercm) [reply src/jobs/release-by-version.yml]: sure

File: README.md [hunk 28]
```diff
@@ -1,2 +1,120 @@
...
+| `is-draft` | boolean | If true, then the release is a draft. | false | `false` |
+| `is-pre-release` | boolean | If true, then the release is a pre-release. | false| `false` |
+
+#### release-by-version
+This is a job to publish a new GitHub Release by checking the version file and CHANGELOG files provided. This is to help ease the process of updating the description for the automated releases. Since this orb is intended for use with Python repositories, the `version-path` parameter defaults to `version.py`.
+
+| Parameter | Type | Description | Required | Default |
+| --- | --- | --- | --- | --- |
+| `version-path` | string | File path to file containing the version. If not specified will default to version.py. We will split the line matching `version=x.y.z` to get the version. | false |  `version.py`|
+| `changelog-path` | string | File path to the changelog containing the description of the release. The format should follow typical [Changlog Symantics](https://keepachangelog.com/en/1.0.0/). | false | `CHANGELOG.md` |
```
Comment (zacharydfriedlander) [inline README.md]: We should also mention that we expect this file to use Markdown?

**Review by zacharydfriedlander — CHANGES_REQUESTED** [review]
One more comment:

File: README.md [hunk 29]
```diff
@@ -1,2 +1,120 @@
...
+* CURRENT_VERSION - The version of the release
+* DESCRIPTION - The description of the release
+* TITLE - The title of the release.
+
+Since this orb is intended for use with Python repositories, the `version-path` parameter defaults to `version.py`.
+
+| Parameter | Type | Description | Required | Default |
+| --- | --- | --- | --- | --- |
+| `version-path` | string | File path to file containing the version. If not specified will default to version.py. We will split the line matching `version=x.y.z` to get the version. | false |  `version.py`|
+| `changelog-path` | string | File path to the changelog containing the description of the release. The format should follow typical Markdown style and [Changlog Symantics](https://keepachangelog.com/en/1.0.0/). | false | `CHANGELOG.md` |
```
Comment (zacharydfriedlander) [inline README.md]: I believe you mean "semantics" here?

File: README.md [hunk 30]
```diff
@@ -1,2 +1,120 @@
...
+| `is-draft` | boolean | If true, then the release is a draft. | false | `false` |
+| `is-pre-release` | boolean | If true, then the release is a pre-release. | false| `false` |
+
+#### release-by-version
+This is a job to publish a new GitHub Release by checking the version file and CHANGELOG files provided. This is to help ease the process of updating the description for the automated releases. Since this orb is intended for use with Python repositories, the `version-path` parameter defaults to `version.py`.
+
+| Parameter | Type | Description | Required | Default |
+| --- | --- | --- | --- | --- |
+| `version-path` | string | File path to file containing the version. If not specified will default to version.py. We will split the line matching `version=x.y.z` to get the version. | false |  `version.py`|
+| `changelog-path` | string | File path to the changelog containing the description of the release. The format should follow typical Markdown style and [Changlog Symantics](https://keepachangelog.com/en/1.0.0/). | false | `CHANGELOG.md` |
```
Comment (zacharydfriedlander) [inline README.md]: Same here.

File: src/jobs/release-by-version.yml [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+  version-path:
+    description: |
+      File path to file containing the version. If not specified will default to version.py.
+      We will split the line matching version=x.y.z to get the version.
+    type: string
+    default: version.py
+  changelog-path:
+    description: |
+      File path to the changelog containing the description of the release.
+      The format should follow typical Markdown style and [Changlog Symantics](https://keepachangelog.com/en/1.0.0/).
```
Comment (zacharydfriedlander) [inline src/jobs/release-by-version.yml]: Flagging "semantics" here as well.

**Review by zacharydfriedlander — CHANGES_REQUESTED** [review]
Overall looking good! One more thing:

**Review by zacharydfriedlander — APPROVED** [review]
LGTM! Great job on all of this!
